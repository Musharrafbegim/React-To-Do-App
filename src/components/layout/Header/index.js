import React from 'react'
import PropTypes from 'prop-types'
import styles from './index.module.scss'

const Header = ({ invert, title, subtitle, rightContent, children }) => {
  return (
    <div className={styles.background} data-test="headerWrapper">
      <div className={styles.container}>
        <div
          className={`${styles.left} ${invert ? styles.invert : ''}`}
          data-test="headerContainer"
        >
          <h1>{title}</h1>
          {!children && <span>{subtitle}</span>}
          {children}
        </div>
        <div className={styles.right} data-test="headerRightContent">
          {rightContent}
        </div>
      </div>
    </div>
  )
}

Header.propTypes = {
  invert: PropTypes.bool,
  title: PropTypes.node.isRequired,
  subtitle: PropTypes.node,
  rightContent: PropTypes.node,
  children: PropTypes.node,
}

Header.defaultProps = {
  invert: false,
  subtitle: '',
  rightContent: null,
  children: null,
}

export default Header
