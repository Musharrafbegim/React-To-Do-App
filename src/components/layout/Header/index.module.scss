@import 'styles/colors.scss';

.background {
  position: relative;
  z-index: 0;
}

.container {
  flex: 1;
  display: flex;
  padding: 2rem 2.1rem 1.4rem;
  z-index: 1;
}

.left {
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  > h1 {
    color: #fff;
    font-size: 2.3rem;
    font-weight: 500;
    margin-top: 9px;
    margin-bottom: 17px;
  }
  > span {
    font-size: 0.95rem;
    color: $outline-color;
    line-height: 30px;
  }
  &.invert {
    align-items: flex-end;
  }
}

.right {
  display: flex;
  align-items: flex-start;
  margin-top: 1.5rem;
}
