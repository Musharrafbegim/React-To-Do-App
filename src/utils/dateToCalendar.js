import { format, isToday, isTomorrow, isThisWeek, isThisYear } from 'date-fns'

export default date => {
  if (isToday(date)) return 'Today'
  if (isTomorrow(date)) return 'Tomorrow'
  if (isThisWeek(date)) return format(date, 'dddd')
  return format(date, `MMM Do${!isThisYear(date) ? ', YYYY' : ''}`)
}
