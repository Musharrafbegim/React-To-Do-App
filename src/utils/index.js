export { default as dateToCalendar } from './dateToCalendar'
export { default as groupByDate } from './groupByDate'
