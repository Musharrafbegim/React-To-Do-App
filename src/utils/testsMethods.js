import checkPropTypes from 'check-prop-types'
import { createStore } from 'redux'
import reducers from 'store/reducers'

export const findByTestAttr = (component, attr) =>
  component.find(`[data-test='${attr}']`)

export const checkProps = ({ propTypes, name }, expectedProps) =>
  checkPropTypes(propTypes, expectedProps, 'props', name)

export const testStore = initialState => createStore(reducers, initialState)
