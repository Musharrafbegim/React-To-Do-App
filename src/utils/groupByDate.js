import dateToCalendar from './dateToCalendar'

export default (list, filterBy) =>
  list.filter(filterBy).reduce((r, acc) => {
    const key = dateToCalendar(acc.date)
    const obj = r
    obj[key] = [...(obj[key] || []), acc]
    return obj
  }, {})
