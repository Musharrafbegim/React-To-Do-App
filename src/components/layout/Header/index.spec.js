import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { findByTestAttr, checkProps } from 'utils/testsMethods'
import Header from '.'

const setUp = (props = {}) => shallow(<Header {...props} />)
const createProps = props => ({
  title: 'Title',
  ...props,
})

describe('Header Component', () => {
  describe('Check PropTypes', () => {
    it('should not throw a warning', () => {
      const expectedProps = {
        title: 'String',
        subtitle: 'String',
        invert: true,
        rightContent: <div>node</div>,
        children: <div>node</div>,
      }

      const propsErr = checkProps(Header, expectedProps)
      expect(propsErr).toBeUndefined()
    })
  })

  describe('Have Props', () => {
    it('should render and matches the snapshot', () => {
      const props = createProps({
        rightContent: <div>node</div>,
        children: <div>node</div>,
      })
      const wrapper = findByTestAttr(setUp(props), 'headerWrapper')

      expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('should render without errors', () => {
      const wrapper = findByTestAttr(setUp(createProps()), 'headerWrapper')

      expect(wrapper.length).toBe(1)
    })

    it('should has invert class if invert is true', () => {
      const props = createProps({ invert: true })
      const headerContainer = findByTestAttr(setUp(props), 'headerContainer')

      expect(headerContainer.hasClass('invert')).toBeTruthy()
    })

    it('should render right content properly', () => {
      const props = createProps({ rightContent: <div>test</div> })
      const headerRightContent = setUp(props).find(
        `[data-test='headerRightContent']`,
      )

      expect(toJson(headerRightContent)).toMatchSnapshot()
    })

    it('should render children if exist as prop', () => {
      const children = <div>test</div>
      const props = createProps({ children })
      const headerContainer = setUp(props).find(`[data-test='headerContainer']`)

      expect(headerContainer.contains(children)).toBeTruthy()
    })

    it("should not render children if doesn't exist as prop", () => {
      const children = <div>test</div>
      const headerContainer = setUp(createProps()).find(
        `[data-test='headerContainer']`,
      )

      expect(headerContainer.contains(children)).toBeFalsy()
    })
  })

  describe('Have NO Props', () => {
    it('should throw a warning', () => {
      const propsErr = checkProps(Header, {})
      expect(propsErr).toBeDefined()
    })
  })
})
