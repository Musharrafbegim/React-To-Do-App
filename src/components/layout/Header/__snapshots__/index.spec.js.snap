// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component Have Props should render and matches the snapshot 1`] = `
<div
  className="background"
  data-test="headerWrapper"
>
  <div
    className="container"
  >
    <div
      className="left "
      data-test="headerContainer"
    >
      <h1>
        Title
      </h1>
      <div>
        node
      </div>
    </div>
    <div
      className="right"
      data-test="headerRightContent"
    >
      <div>
        node
      </div>
    </div>
  </div>
</div>
`;

exports[`Header Component Have Props should render right content properly 1`] = `
<div
  className="right"
  data-test="headerRightContent"
>
  <div>
    test
  </div>
</div>
`;
